### PR TITLE
WIP: Connection Request and Response

### PR DESCRIPTION
**Pull request by mdwaud**
Implements https://github.com/sparkmeter/sink/issues/3 except for communicating which event type ids and schemas are supported.
- [x] encoding / decoding of connection request and connection response
- [x] implement protocol for `{:connection_response, :ok}`
- [x] implement protocol for `{:connection_response, {:hello_new_client, _}}`
- [x] implement protocol for `{:connection_response, {:mismatched_client, _}}`
- [x] implement protocol for `{:connection_response, {:mismatched_server, _}}`
- [x] implement protocol for `{:connection_response, {:quarantined, _, _}}`
- [x] implement protocol for `{:connection_response, :unsupported_protocol_version}`
- [x] implement protocol for `{:connection_response, :unsupported_application_version}`

### CODE REVIEW

**Comment by mdwaud** [comment]
Almost there. Just need to implement unquarantining a client (or punt on that for a future PR) and do some cleanup.

File: lib/sink/connection/server_handler.ex [hunk 1]
```diff
@@ -316,12 +360,57 @@ defmodule Sink.Connection.ServerHandler do
...
+        {:error, :unsupported_protocol_version, protocol_version} ->
+          result = {:unsupported_protocol_version, protocol_version}
+          frame = Connection.Protocol.encode_frame({:connection_response, result})
+
+          new_connection_status =
+            ConnectionStatus.connection_request(state.connection_status, result)
+
+          new_state = %State{state | connection_status: new_connection_status}
+          handler.handle_connection_response(new_state.client, result)
+          # todo: close connection
```
Comment (mdwaud) [inline lib/sink/connection/server_handler.ex:380]: Can remove this todo now that we close the connection

File: lib/sink/connection/server_handler.ex [hunk 2]
```diff
@@ -411,6 +500,8 @@ defmodule Sink.Connection.ServerHandler do
         :pong ->
           Sink.Telemetry.pong(:received, %{client_id: client_id})
           {state, nil}
+
+          # todo: handle events when connection is not active
```
Comment (mdwaud) [inline lib/sink/connection/server_handler.ex:504]: todo can be removed

File: lib/sink/connection/client/connection_status.ex [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+  @type t :: %__MODULE__{
+          connection_state: :requesting_connection | :connected | :disconnecting,
+          server_identifier: Protocol.server_identifierentifier() | nil,
+          reason: nil | binary
+        }
+
+  defstruct [
+    :connection_state,
+    :server_identifier,
```
Comment (mdwaud) [inline lib/sink/connection/client/connection_status.ex:15]: How about `instance_id`? That should be generic enough and implies that the identifier is tied to the instance of the pairing of client and server vs just the server.

**Comment by mdwaud** [comment]
Looks good, thanks for taking this over. A few small comments / questions, but otherwise looks great. The property testing is a nice touch.

File: lib/sink/connection/client/connection_status.ex [hunk 4]
```diff
@@ -0,0 +1,72 @@
+defmodule Sink.Connection.Client.ConnectionStatus do
+  @moduledoc """
+  State machine that manages a connection's state and state transitions
+  """
+  alias Sink.Connection.Protocol
+
+  @type t :: %__MODULE__{
+          connection_state: :requesting_connection | :connected | :disconnecting,
+          server_identifier: Protocol.server_identifierentifier() | nil,
```
[the remark this answers is not in the document]
Comment (LostKobrakai) [reply lib/sink/connection/client/connection_status.ex:9]: Good old search and replace issue I didn't correct :D

File: lib/sink/connection/protocol.ex [hunk 5]
```diff
@@ -2,47 +2,137 @@ defmodule Sink.Connection.Protocol do
...
+          | {:ack, message_id}
+          | {:publish, message_id, payload :: binary}
+          | :ping
+          | :pong
+          | {:nack, message_id, payload :: binary}
+
+  @type nack_data() :: {binary(), String.t()}
+
+  # when we get to Sink 1.0 this will be 0 and we'll delete / deprecate anything
+  # 8 and higher
```
Comment (LostKobrakai) [inline lib/sink/connection/protocol.ex:32]: @mdwaud why use `8` here? protocol version is encoded using 4 bytes, so `0..255`, so it feels a bit waistful to deprecate anything 8 or higher.

**Pull request merged**